OCPBUGS-9133: pkg/cvo/metrics: Connect ClusterVersion to ClusterOperatorDown and ClusterOperatorDegraded

wking opened the pull request:
By adding `cluster_operator_up` handling for ClusterVersion, with `version` as the component name, the same way we handle `cluster_operator_conditions`.  This plugs us into `ClusterOperatorDown` (based on `cluster_operator_up`) and `ClusterOperatorDegraded` (based on both `cluster_operator_conditions` and `cluster_operator_up`).

I've adjusted the `ClusterOperatorDegraded` rule so that it fires on ClusterVersion `Failing=True` and does not fire on `Failing=False`.  Thinking through an update from before:

1. Outgoing CVO does not serve `cluster_operator_up{name="version"}`.
2. User requests an update to a release with this change.
3. New CVO comes in, starts serving `cluster_operator_up{name="version"}`.
4. Old `ClusterOperatorDegraded` no matching `cluster_operator_conditions{name="version",condition="Degraded"}`, falls through to `cluster_operator_up{name="version"}`, and starts cooking the `for: 30m`.
5. If we go more than 30m before updating the `ClusterOperatorDegraded` rule to understand `Failing`, `ClusterOperatorDegraded` would fire.

We'll need to backport the `ClusterOperatorDegraded` `expr` change to one 4.y release before the CVO-metrics change lands to get:

1. Outgoing CVO does not serve `cluster_operator_up{name="version"}`.
2. User requests an update to a release with the `expr` change.
3. Incoming `ClusterOperatorDegraded` sees no `cluster_operator_conditions{name="version",condition="Degraded"}`, `cluster_operator_conditions{name="version",condition="Failing"}` (we hope), or `cluster_operator_up{name="version"}`, so it doesn't fire.  Unless we are `Failing=True`, in which case, hooray, we'll start alerting about it.
4. User requests an update to a release with the CVO-metrics change.
5. New CVO starts serving `cluster_operator_up{name="version"}`, just like the fresh-modern-install situation, and everything is great.

The missing-ClusterVersion metrics don't matter all that much today, because the CVO has been creating replacement ClusterVersion since at least 90e9881bd5 (#45).  But it will become more important with #741, which is planning on removing that default creation.  When there is no ClusterVersion, we expect `ClusterOperatorDown` to fire.

wking commented:
/retitle OCPBUGS-9133: pkg/cvo/metrics: Connect ClusterVersion to ClusterOperatorDown and ClusterOperatorDegraded

wking commented:
/jira refresh

wking commented:
/remove-lifecycle rotten

wking commented:
The `machine-os-content` business has been sorted out.

/retest-required

wking commented:
lots of `Pod scheduling timeout`:

/retest-required

wking commented:
[e2e-agnostic-operator][1] failed to bootstrap, with the bootstrap CVO panicking:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/746/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic-operator/1772845119326654464/artifacts/e2e-agnostic-operator/ipi-install-install/artifacts/log-bundle-20240327053130.tar | tar -xOz log-bundle-20240327053130/bootstrap/containers/cluster-version-operator-654ef1fecd6ccd9d19174878a564063c31267eb2a50a99e6afaf175ebf664815.log | grep -B1 -A100 'Observed a panic:'
I0327 05:30:33.831324       1 cvo.go:474] Collected cluster version sync goroutine.
E0327 05:30:33.831292       1 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 175 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x1bba1c0?, 0x30c9c80})
        /go/src/github.com/openshift/cluster-version-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:75 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x0?})
        /go/src/github.com/openshift/cluster-version-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49 +0x6b
panic({0x1bba1c0?, 0x30c9c80?})
        /usr/lib/golang/src/runtime/panic.go:914 +0x21f
github.com/openshift/cluster-version-operator/pkg/cvo.(*Operator).Run.func1()
        /go/src/github.com/openshift/cluster-version-operator/pkg/cvo/cvo.go:416 +0x6f
created by github.com/openshift/cluster-version-operator/pkg/cvo.(*Operator).Run in goroutine 89
        /go/src/github.com/openshift/cluster-version-operator/pkg/cvo/cvo.go:414 +0x40f
I0327 05:30:33.831328       1 cvo.go:468] Waiting on 3 outstanding goroutines.
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x19420af]

goroutine 175 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x0?})
        /go/src/github.com/openshift/cluster-version-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:56 +0xcd
panic({0x1bba1c0?, 0x30c9c80?})
        /usr/lib/golang/src/runtime/panic.go:914 +0x21f
github.com/openshift/cluster-version-operator/pkg/cvo.(*Operator).Run.func1()
        /go/src/github.com/openshift/cluster-version-operator/pkg/cvo/cvo.go:416 +0x6f
created by github.com/openshift/cluster-version-operator/pkg/cvo.(*Operator).Run in goroutine 89
        /go/src/github.com/openshift/cluster-version-operator/pkg/cvo/cvo.go:414 +0x40f
```

Looks like [that is][2]:

```go
	// trigger the first cluster version reconcile always
	optr.queue.Add(optr.queueKey())

	resultChannelCount++
	go func() {
		defer utilruntime.HandleCrash()
panic!		runThrottledStatusNotifier(runContext, optr.statusInterval, 2, optr.configSync.StatusCh(), func() { optr.queue.Add(optr.queueKey()) })
		resultChannel <- asyncResult{name: "status notifier"}
	}()
```

With `optr.queue.Add(optr.queueKey())` completing successfully a few lines earlier, and [`statusInterval` a non-pointer `Duration`][3], I'd guess that the `optr.configSync.StatusCh()` call is the issue.  That gets [set up in `InitializeFromPayload`][4], which [is called before `Operator.Run`][5], so I'm not clear on how we're panicking there.  But possibly worth digging more in a separate pull, since it seems orthogonal to my changes.

/retest

[1]: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/746/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic-operator/1772845119326654464
[2]: https://github.com/openshift/cluster-version-operator/blame/8f4a120d70e626496fcfec5f877ec0f290e9af92/pkg/cvo/cvo.go#L416
[3]: https://github.com/openshift/cluster-version-operator/blame/8f4a120d70e626496fcfec5f877ec0f290e9af92/pkg/cvo/cvo.go#L157
[4]: https://github.com/openshift/cluster-version-operator/blame/8f4a120d70e626496fcfec5f877ec0f290e9af92/pkg/cvo/cvo.go#L319
[5]: https://github.com/openshift/cluster-version-operator/blame/8f4a120d70e626496fcfec5f877ec0f290e9af92/pkg/start/start.go#L271-L282

wking commented:
Aha, here's why [CI runs][1] are failing:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/746/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic-ovn/1777506846186147840/artifacts/e2e-agnostic-ovn/ipi-install-install/artifacts/log-bundle-20240409022023.tar | tar -xOz log-bundle-20240409022023/bootstrap/containers/cluster-version-operator-d757dec43012ea3097967701b53f4c31cdf7255212882a43327491402da189d4.log | grep -A1 'Failed to initialize from payload'
I0409 02:13:08.579037       1 start.go:250] Failed to initialize from payload; shutting down: the local release contents are invalid - no current version can be determined from disk: Error loading manifests from /: preprocess 0000_90_cluster-version-operator_02_servicemonitor.yaml: failed to parse manifest: template: manifest:99: undefined variable "$labels"
I0409 02:13:08.579062       1 start.go:254] Launching CVO.Run gorountine
```

Looks like we're calling `Run` there when we should be shutting down...

[1]: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/746/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic-ovn/1777506846186147840

dis016 commented:
Hi @wking, i am unable to spin up the cluster due to above bootstrap error, https://prow.ci.openshift.org/view/gs/test-platform-results/logs/release-openshift-origin-installer-launch-aws-modern/1777295509413695488

wking commented:
20ec9573 -> 74312d17 drops some debug logging, now that I understand why the panic was happening.  And it  adds a second layer of template quoting to the alerting-rule `description` strings, to hopefully avoid the `undefined variable "$labels"` issue.

wking commented:
@dis016 , CI success like [this][1] suggests I have sorted out the `undefined variable "$labels"` issue, if you wanted to try pre-merge testing again :)

/retest required

[1]: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-version-operator/746/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic-ovn/1777582448440250368

wking commented:
/retest-required

wking commented:
Testing with Cluster Bot and `launch 4.16,openshift/cluster-version-operator#746 aws` ([logs][1]), I opened by making auth mad:

```console
$ oc adm cordon -l node-role.kubernetes.io/control-plane=
node/ip-10-0-120-3.us-west-1.compute.internal cordoned
node/ip-10-0-31-66.us-west-1.compute.internal cordoned
node/ip-10-0-71-7.us-west-1.compute.internal cordoned
$ oc -n openshift-authentication delete "$(oc -n openshift-authentication get -o name pods | head -n1)"
pod "oauth-openshift-6cbcb6579f-2fg22" deleted
```

Do the Machine-approver too, for good measure, as I pick on things that should spook cluster-operators without actually hurting cluster performance (I'm not trying to scale up new Machines/Nodes):

```console
$ oc -n openshift-cluster-machine-approver delete "$(oc -n openshift-cluster-machine-approver get -o name pods | head -n1)"
pod "machine-approver-74b7866855-z2flp" deleted
```

With operand pods removed, and the cordon blocking its replacement from scheduling, the operators should be grumbling, but even after 15m, they're still all happy:

```console
$ oc get -o json clusteroperator | jq -c '.items[].status.conditions[] | select(.type == "Available" or .type == "Degraded") | {type, status}' | sort | uniq -c
     33 {"type":"Available","status":"True"}
     33 {"type":"Degraded","status":"False"}
```

Maybe I should go after the registry:

```console
$ oc adm cordon -l node-role.kubernetes.io/worker=
node/ip-10-0-108-53.us-west-1.compute.internal cordoned
node/ip-10-0-33-86.us-west-1.compute.internal cordoned
node/ip-10-0-87-58.us-west-1.compute.internal cordoned
$ oc delete namespace openshift-image-registry
namespace "openshift-image-registry" deleted
```

Hey, now an operator is mad:

```console
$ oc get -o json clusteroperator | jq -c '.items[] | .metadata.name as $n | .status.conditions[] | select((.type == "Available" and .status == "False") or (.type == "Degraded" and .status == "True")) | .name = $n' | sort
{"lastTransitionTime":"2024-04-10T06:48:00Z","message":"1 of 6 credentials requests are failing to sync.","reason":"CredentialsFailing","status":"True","type":"Degraded","name":"cloud-credential"}
```

But that's not one I'd been trying to poke, and then it got happy again.  Probably just trying to create the registry's CredentialsRequest Secret, and struggling until the CVO had recreated that namespace.  Ah, eventually `machine-config` complains about the cordoned control plane:

```console
$ oc get -o json clusteroperator | jq -c '.items[] | .metadata.name as $n | .status.conditions[] | select((.type == "Available" and .status == "False") or (.type == "Degraded" and .status == "True")) | .name = $n' | sort
{"lastTransitionTime":"2024-04-10T06:50:28Z","message":"Failed to resync 4.16.0-0.test-2024-04-10-052103-ci-ln-xtnbb9k-latest because: error during syncRequiredMachineConfigPools: [context deadline exceeded, failed to update clusteroperator: [client rate limiter Wait returned an error: context deadline exceeded, error required MachineConfigPool master is not ready, retrying. Status: (total: 3, ready 0, updated: 3, unavailable: 3, degraded: 0)]]","reason":"RequiredPoolsFailed","status":"True","type":"Degraded","name":"machine-config"}
```

and ~5m later, `authentication` starts complaining too:

```console
$ oc get -o json clusteroperator | jq -c '.items[] | .metadata.name as $n | .status.conditions[] | select((.type == "Available" and .status == "False") or (.type == "Degraded" and .status == "True")) | .name = $n' | sort
{"lastTransitionTime":"2024-04-10T06:50:28Z","message":"Failed to resync 4.16.0-0.test-2024-04-10-052103-ci-ln-xtnbb9k-latest because: error during syncRequiredMachineConfigPools: [context deadline exceeded, failed to update clusteroperator: [client rate limiter Wait returned an error: context deadline exceeded, error required MachineConfigPool master is not ready, retrying. Status: (total: 3, ready 0, updated: 3, unavailable: 3, degraded: 0)]]","reason":"RequiredPoolsFailed","status":"True","type":"Degraded","name":"machine-config"}
{"lastTransitionTime":"2024-04-10T06:55:30Z","message":"OAuthServerDeploymentDegraded: 1 of 3 requested instances are unavailable for oauth-openshift.openshift-authentication ()","reason":"OAuthServerDeployment_UnavailablePod","status":"True","type":"Degraded","name":"authentication"}
```

And the CVO passes these along:

```console
$ oc adm upgrade
Failing=True:

  Reason: ClusterOperatorsDegraded
  Message: Cluster operators authentication, machine-config are degraded

Error while reconciling 4.16.0-0.test-2024-04-10-052103-ci-ln-xtnbb9k-latest: authentication, machine-config has an unknown error: ClusterOperatorsDegraded
...
```

`version` alert renders well:

![image](https://github.com/openshift/cluster-version-operator/assets/209920/52fb6100-8082-4d39-b983-066b6fa31f94)

operator alert renders with an extra `}}`:

![image](https://github.com/openshift/cluster-version-operator/assets/209920/73ab9533-ec81-495f-8559-361d6d23193d)

[1]: https://prow.ci.openshift.org/view/gs/test-platform-results/logs/release-openshift-origin-installer-launch-aws-modern/1777927392308760576

wking commented:
[earlier testing][1]:
> operator alert renders with an extra `}}`:

I've pushed 74312d17 -> 10849d7c to address this.

[1]: https://github.com/openshift/cluster-version-operator/pull/746#issuecomment-2046669724

petr-muller commented on pkg/start/start.go:
```diff
@@ -247,6 +247,7 @@ func (o *Options) run(ctx context.Context, controllerCtx *Context, lock resource
 						}
 						klog.Infof("Failed to initialize from payload; shutting down: %v", err)
 						resultChannel <- asyncResult{name: "payload initialization", error: firstError}
+						return
```
This was the panic?

wking replied:
yup, essay in 2952a2fe6990f1a2fe5e ;)

dis016 commented:
/qe-approved

jiajliu commented:
/label qe-approved cc @dis016

wking commented:
/label qe-approved
